### PR TITLE
Add RootChecked to guarantee root node type

### PR DESCRIPTION
**Pull request by acl-cqc**
* Factor out RootTagged as a subtrait of HugrView to allow separate implementation. Most things can still just use HugrView (as it happens everything that implements HugrView also implements RootTagged, but this is not required).
    * HugrMut a subtrait of RootTagged not just HugrView
* Then add a new RootChecked struct storing any `AsRef<Hugr>` with a fixed RootHandle.
     * It allows `.as_ref()`, throwing away the extra information in the RootHandle
     * If the underlying view is a HugrMut then it is too
     * Use trait-default implementations of Hugr(View,(Mut)Internals), as the latter check that `replace_op` does not break the bound on root-type
     * But do not provide `.as_mut()` as that would allow bypassing and invalidating the extra info in the RootHandle
     * Like SiblingMut, check that nested instances only narrow the bound.
* Change the overridden-for-&(mut) Hugr impls of Hugr(Mut)Internals to only work for `RootHandle=Node` to ensure the lack of checking in replace_op there is safe.

### CODE REVIEW

**Review by ss2165 — COMMENTED** [review]
Seems fine to me, but I would suggest waiting for Agustin's review too

File: src/hugr/views.rs [hunk 1]
```diff
@@ -303,6 +305,29 @@ pub trait HugrView: sealed::HugrInternals {
...
 
+/// A view of the whole Hugr.
+/// (Just provides static checking of the type of the root node)
+pub struct WholeHugrView<'a, Root = Node>(&'a Hugr, PhantomData<Root>);
+
+impl<'a, Root: NodeHandle> WholeHugrView<'a, Root> {
+    /// Create a hierarchical view of a whole HUGR
+    ///
+    /// # Errors
+    /// Returns [`HugrError::InvalidNode`] if the root isn't a node of the required [OpTag]
```
Comment (aborgna-q) [inline src/hugr/views.rs:316]: I feel we could use a `HugrError::InvalidTag` variant for these cases. This could be another PR, since we do the same on all the views.

Comment (aborgna-q) [inline src/hugr/views.rs:316]:
```suggestion
    /// Returns [`HugrError::InvalidNode`] if the root isn't a node of the required [`OpTag`]
    ///
    /// [`OpTag`]: crate::ops::tag::OpTag
```

File: src/hugr/views.rs [hunk 2]
```diff
@@ -303,6 +305,29 @@ pub trait HugrView: sealed::HugrInternals {
...
+    pub fn try_new(hugr: &'a Hugr) -> Result<Self, HugrError> {
```
Comment (aborgna-q) [inline src/hugr/views.rs:317]: impl `TryInto<WholeHugrView<'a, R>> for &'a Hugr` ?

**Review by aborgna-q — COMMENTED** [review]
I agree with doing root-checked references, but from knowing how the rest of the codebase went we will also need a mutable versions at some point.

What do you think about having a
```rust
pub struct RootTagged<H, Root>(H, PhantomData<Root>);

impl<H: AsRef<Hugr>, Root: NodeHandle> RootTagged<H, Root> { ... }
```
instead? (and derive `AsMut` too).

File: src/hugr/views.rs (hunk 1 above)
Comment (acl-cqc) [reply src/hugr/views.rs]: Done (the extra link definition didn't seem to be required)

File: src/hugr/views.rs (hunk 2 above)
Comment (acl-cqc) [reply src/hugr/views.rs]: Nice!! :-)

**Comment by acl-cqc** [comment]
> I agree with doing root-checked references, but from knowing how the rest of the codebase went we will also need a mutable versions at some point.
> 
> What do you think about having....
> 
> ```rust
> impl<H: AsRef<Hugr>, Root: NodeHandle> RootTagged<H, Root> { ... }
> ```
> 
> instead? (and derive `AsMut` too).

Not sure I understand the lifetime stuff well enough, but don't we need the explicit `'a` lifetime to make this useful?

Hmmm, I'll have a play and see where I get. `AsMut` (or `HugrMut` for the `WholeHugrView` as it stands) both seem reasonable, yes can do (but will try to figure out the AsRef-vs-`'a` before I do that)

**Comment by acl-cqc** [comment]
> don't we need the explicit `'a` lifetime to make this useful?

Hmmm, probably not. The likely use case here is when you have an (&)Hugr and want to pass it to something that takes an `&impl HugrView<NodeHandle=...>`, see for example #247. Returning a WholeHugrView with a particular lifetime specified in the type is...possible, but likely not the main use case.

So yeah, maybe we don't need the lifetime after all....but, that said, what's the advantage of using `AsRef<Hugr>`? You gain the ability to store a whole Hugr in a WholeHugrView, if you want to. In general I doubt that's the main use case *either*, but again, probably has uses in other/different corner cases...

File: src/hugr/views.rs (hunk 1 above)
Comment (acl-cqc) [reply src/hugr/views.rs]: Yes, #581, I think it'd be good to do that first, and then I'll make this PR update those tests

**Comment by acl-cqc** [comment]
> I agree with doing root-checked references, but from knowing how the rest of the codebase went we will also need a mutable versions at some point.

Yup, now done :).

> What do you think about having a
> 
> ```rust
> pub struct RootTagged<H, Root>(H, PhantomData<Root>);
> ```

Ok, I persuade myself that the use case of returning a `RootTagged<Hugr, DfgId>` (owning the Hugr) is useful - otherwise you would need to return a Hugr and a WholeHugrView of that same thing, which is generally impossible. So, good call.

That said, it's a shame to see `try_from` has to go, and it's a bit of a pain having to do `RootTagged<Hugr, XXXId>` when you want to specify just the XXX. But the only way around that that I can see is something like:
```rust
struct RootTagFactory<H>(PhantomData<H>);
impl<H: AsRef<Hugr>> RootTagFactory<H> {
  fn try_new<Root: NodeHandle>(h: H) -> Result<RootTagged<H, Root>> {....}
}
```
so you can then do `RootTagFactory::try_new::<CfgID>(h)`. (I think? Right?)....I am not sure that's worth it.

File: src/hugr/hugrmut.rs [hunk 3]
```diff
@@ -216,10 +216,7 @@ impl InsertionResult {
 }
 
 /// Impl for non-wrapped Hugrs. Overwrites the recursive default-impls to directly use the hugr.
-impl<T> HugrMut for T
-where
-    T: HugrView + AsMut<Hugr>,
-{
+impl<T: HugrView + AsMut<Hugr>> HugrMut for T {
```
Comment (acl-cqc) [inline src/hugr/hugrmut.rs:219]: Hmmm, these were a bit drive-by, but they make searching for these quasi-blanket impls much easier. (A one-line search hit `impl<T> HugrMut for T` is really not very informative!)

**Comment by acl-cqc** [comment]
CI failing due to updates in clippy. Let's fix those in another PR.

**Comment by acl-cqc** [comment]
> CI failing on existing code due to updates in clippy. Let's fix those in another PR.

Done (#592)

File: src/hugr/views.rs [hunk 4]
```diff
@@ -303,8 +300,59 @@ pub trait HugrView: sealed::HugrInternals {
...
+    pub fn into_hugr(self) -> Hugr {
+        self.0
+    }
+}
+
+impl<H: AsRef<Hugr>, Root: NodeHandle> RootTagged for RootChecked<H, Root> {
+    type RootHandle = Root;
+}
+
+// Note do not implement AsMut<Hugr> - that would get us the `impl HugrMutInternals`
```
Comment (acl-cqc) [inline src/hugr/views.rs:338]: Note this contrary to your initial idea @aborgna-q. If you try to impl AsMut<Hugr>, you'll have to remove the impl of HugrMutInternals (and HugrMut) here, and then the test will fail. The test may be enough to justify removing this comment, possibly.

File: src/hugr/views.rs [hunk 5]
```diff
@@ -451,10 +508,7 @@ pub(crate) mod sealed {
         fn root_node(&self) -> Node;
     }
 
-    impl<T> HugrInternals for T
-    where
-        T: AsRef<super::Hugr>,
-    {
+    impl<T: AsRef<super::Hugr>> HugrInternals for T {
```
Comment (aborgna-q) [inline src/hugr/views.rs:511]:
```suggestion
    impl<T: AsRef<Hugr>> HugrInternals for T {
```

File: src/hugr/views.rs [hunk 6]
```diff
@@ -303,8 +300,61 @@ pub trait HugrView: sealed::HugrInternals {
...
+    /// The kind of handle that can be used to refer to the root node.
+    ///
+    /// The handle is guaranteed to be able to contain the operation returned by
+    /// [`HugrView::root_type`].
+    type RootHandle: NodeHandle;
+}
+
+/// A view of the whole Hugr.
+/// (Just provides static checking of the type of the root node)
+pub struct RootChecked<H, Root = Node>(H, PhantomData<Root>);
```
Comment (aborgna-q) [inline src/hugr/views.rs:314]: This should go in a submodule, as all the other views.

File: src/hugr/views.rs (hunk 4 above)
Comment (aborgna-q) [reply src/hugr/views.rs:338]: Uhm, this is an easy pit to fall in.

Since the auto-impl for `HugrMutInternals` requires `RootTagged`, why not check the root type before dereferencing with `as_mut` there?
The generic constraints in `impl<T: RootTagged + AsMut<Hugr>> HugrMutInts for T` kind of tell you that you're adding the extra check on top of the reference.

Comment (acl-cqc) [reply src/hugr/views.rs]: Hmmm, the answer might be - premature optimization, i.e. that the `impl HugrMutInternals` for `T: AsMut<Hugr>` then gets to skip the check....so ok, let's not prematurely optimize, SGTM ;)

Note there was an unexpected subtle consequence that I had to rule out in 07e4f1097cf936d8327881bdc0e2b118ba627613

Comment (acl-cqc) [reply src/hugr/views.rs]: (and this does let anyone call `.as_mut().replace_op(...)` and bypass the check, which I'm not very keen on....)

Comment (acl-cqc) [reply src/hugr/views.rs]: So overall, I have done as you say, but I'm not sure why this is better, TBH

Comment (acl-cqc) [reply src/hugr/views.rs]: See 2be330dc7a16dccc7dacbf04bbe84c63fa5e283c, which HEAD reverts - I actually think I prefer the former; it needs an extra `impl HugrMut` (but, empty!) so that seems minimally harmful to me

**Review by aborgna-q — APPROVED** [review]
👍

**Comment by acl-cqc** [comment]
I'm now allowing making one RootChecked on top of another, or indeed on top of a SiblingMut, so long as the root-type bound is only narrowed, rather than only on top of something whose RootHandle was Any, as before. And hiding `.as_mut()` (which would allow breaking the bound) by using the trait-default HugrMut methods rather than the versions overridden for operating directly on a Hugr.
